### PR TITLE
Homework OOP Advanced

### PR DESCRIPTION
**Pull request by vojtcekhovski**

### CODE REVIEW

File: src/main/java/core/basesyntax/FigureArea.java [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package core.basesyntax;
+
+interface FigureArea {
```
Comment (zachar-chorny) [inline src/main/java/core/basesyntax/FigureArea.java:3]:
```suggestion
public interface AreaCalculator {
```

File: src/main/java/core/basesyntax/FigureDrawable.java [hunk 2]
```diff
@@ -0,0 +1,5 @@
+package core.basesyntax;
+
+public interface FigureDrawable {
```
Comment (zachar-chorny) [inline src/main/java/core/basesyntax/FigureDrawable.java:3]:
```suggestion
public interface Drawable {

```

File: src/main/java/core/basesyntax/FigureSupplier.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
+package core.basesyntax;
+
+import java.util.Random;
+
+public class FigureSupplier {
+    public Figure getRandomFigure() {
+        Random random = new Random();
+        ColorSupplier colorSupplier = new ColorSupplier();
```
Comment (zachar-chorny) [inline src/main/java/core/basesyntax/FigureSupplier.java:8]: make these variables a class-fields, like in your ColorSupplier

File: src/main/java/core/basesyntax/FigureSupplier.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+                return new RightTriangle(colorSupplier.getRandomColor(),
+                        random.nextInt(100),
+                        random.nextInt(100));
+            case (4):
+                return new Square(colorSupplier.getRandomColor(),
+                        random.nextInt(100));
+            default:
+                return getDefaultFigure();
+
+        }
```
Comment (zachar-chorny) [inline src/main/java/core/basesyntax/FigureSupplier.java:32]: lets create constants for magic numbers(5, 100)

File: src/main/java/core/basesyntax/FigureSupplier.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    public Figure getDefaultFigure() {
+        return new Circle(Color.WHITE, 10);
```
Comment (zachar-chorny) [inline src/main/java/core/basesyntax/FigureSupplier.java:36]: also create constants for White and 10

File: src/main/java/core/basesyntax/FigureSupplier.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+                        random.nextInt(100));
+            case (3):
+                return new RightTriangle(colorSupplier.getRandomColor(),
+                        random.nextInt(100),
+                        random.nextInt(100));
+            case (4):
+                return new Square(colorSupplier.getRandomColor(),
+                        random.nextInt(100));
+            default:
+                return getDefaultFigure();
```
Comment (zachar-chorny) [inline src/main/java/core/basesyntax/FigureSupplier.java:30]: create private methods to create each figure and use them here

File: src/main/java/core/basesyntax/FigureSupplier.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
```
Comment (zachar-chorny) [inline src/main/java/core/basesyntax/FigureSupplier.java:31]: empty line

File: src/main/java/core/basesyntax/Main.java [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+    public static void main(String[] args) {
+        FigureSupplier figureSupplier = new FigureSupplier();
+
+        Figure[] figures = new Figure[6];
+        figures[0] = figureSupplier.getRandomFigure();
+        figures[1] = figureSupplier.getRandomFigure();
+        figures[2] = figureSupplier.getRandomFigure();
+        figures[3] = figureSupplier.getDefaultFigure();
+        figures[4] = figureSupplier.getDefaultFigure();
+        figures[5] = figureSupplier.getDefaultFigure();
```
Comment (zachar-chorny) [inline src/main/java/core/basesyntax/Main.java:13]: lets use loop for it

**Review by zachar-chorny — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/FigureSupplier.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
+package core.basesyntax;
+
+import java.util.Random;
+
+public class FigureSupplier {
+    private final Random random = new Random(System.currentTimeMillis());
```
Comment (aleksandr-bevza) [inline src/main/java/core/basesyntax/FigureSupplier.java:6]: I think this is not needed
```suggestion
    private final Random random = new Random();

```

File: src/main/java/core/basesyntax/FigureSupplier.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
+package core.basesyntax;
+
+import java.util.Random;
+
+public class FigureSupplier {
+    private final Random random = new Random(System.currentTimeMillis());
+    private final ColorSupplier colorSupplier = new ColorSupplier();
+    private final int randomMaxValue = 100;
+    private final int defaultRadius = 10;
```
Comment (aleksandr-bevza) [inline src/main/java/core/basesyntax/FigureSupplier.java:9]: Please follow the rules of constants naming
```suggestion
    private static final int RANDOM_MAX_VALUE = 100;
    private static final int DEFAULT_RADIUS = 10;
```

File: src/main/java/core/basesyntax/FigureSupplier.java [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    private final ColorSupplier colorSupplier = new ColorSupplier();
+    private final int randomMaxValue = 100;
+    private final int defaultRadius = 10;
+
+    public final Figure getDefaultFigure() {
+        Color defaultColor = Color.WHITE;
+        return new Circle(defaultColor, defaultRadius);
+    }
+
+    public Figure getRandomFigure() {
```
Comment (aleksandr-bevza) [inline src/main/java/core/basesyntax/FigureSupplier.java:16]: Not fixed. Please create private methods for each figure creation.
image

File: src/main/java/core/basesyntax/FigureSupplier.java [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+import java.util.Random;
+
+public class FigureSupplier {
+    private final Random random = new Random(System.currentTimeMillis());
+    private final ColorSupplier colorSupplier = new ColorSupplier();
+    private final int randomMaxValue = 100;
+    private final int defaultRadius = 10;
+
+    public final Figure getDefaultFigure() {
+        Color defaultColor = Color.WHITE;
```
Comment (aleksandr-bevza) [inline src/main/java/core/basesyntax/FigureSupplier.java:12]: Create default color as constant, move it to class field:
private static final Color DEFAULT_COLOR = Color.WHITE;

**Review by aleksandr-bevza — CHANGES_REQUESTED** [review]

**Comment by vojtcekhovski** [comment]
Create default color as constant and create private methods for each figure

**Review by vojtcekhovski — COMMENTED** [review]
Create default color as constant and create private methods for each figure

File: src/main/java/core/basesyntax/FigureSupplier.java [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+
+    public final Figure getDefaultFigure() {
+        return new Circle(DEFAULT_COLOR, DEFAULT_RADIUS);
+    }
+
+    private Circle getRandomCircle() {
+        return new Circle(colorSupplier.getRandomColor(),
+                random.nextInt(RANDOM_MAX_VALUE));
+    }
+
```
Comment (okuzan) [inline src/main/java/core/basesyntax/FigureSupplier.java:20]: private members should go after public ones

File: src/main/java/core/basesyntax/Main.java [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+            if (i < figures.length / 2) {
+                figures[i] = figureSupplier.getRandomFigure();
+            } else {
+                figures[i] = figureSupplier.getDefaultFigure();
+            }
+        }
+        
+        for (Figure figure : figures) {
+            figure.drawFigure();
+        }
```
Comment (okuzan) [inline src/main/java/core/basesyntax/Main.java:18]: no need to iterate twice, you can call it in the first loop

File: src/main/java/core/basesyntax/IsoscelesTrapezoid.java [hunk 15]
```diff
@@ -0,0 +1,24 @@
+package core.basesyntax;
+
+public class IsoscelesTrapezoid extends Figure {
+    private double median;
```
Comment (okuzan) [inline src/main/java/core/basesyntax/IsoscelesTrapezoid.java:4]: better include two parallel sides, the median can be calculated based on them. How we don't have complete info about a figure

File: src/main/java/core/basesyntax/Drawable.java [hunk 16]
```diff
@@ -0,0 +1,5 @@
+package core.basesyntax;
+
+public interface Drawable {
+    void drawFigure();
```
Comment (okuzan) [inline src/main/java/core/basesyntax/Drawable.java:4]:
```suggestion
    void draw();

```

**Review by okuzan — CHANGES_REQUESTED** [review]

**Review by vojtcekhovski — COMMENTED** [review]
include two parallel sides, private members go after public ones

File: src/main/java/core/basesyntax/Main.java [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+        FigureSupplier figureSupplier = new FigureSupplier();
+        Figure[] figures = new Figure[6];
+
+        for (int i = 0; i < figures.length; i++) {
+            if (i < figures.length / 2) {
+                figures[i] = figureSupplier.getRandomFigure();
+                figures[i].draw();
+            } else {
+                figures[i] = figureSupplier.getDefaultFigure();
+                figures[i].draw();
```
Comment (olekskov) [inline src/main/java/core/basesyntax/Main.java:14]: .draw() method called in any case so should be extracted from condition block

**Review by olekskov — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/Main.java [hunk 18]
```diff
@@ -0,0 +1,18 @@
...
+        Figure[] figures = new Figure[6];
+
+        for (int i = 0; i < figures.length; i++) {
+            if (i < figures.length / 2) {
+                figures[i] = figureSupplier.getRandomFigure();
+                System.out.println("Random figure: " + figures[i]);
+            } else {
+                figures[i] = figureSupplier.getDefaultFigure();
+                System.out.println("Default figure: " + figures[i]);
+            }
```
Comment (olekskov) [inline src/main/java/core/basesyntax/Main.java:15]: Sorry, maybe I wasn't clear enough. Previous version with .draw() method usage was correct in general.
What were required to do is just extract it from conditions to use one time. Like:
if (condition) {
    figures[I] = ...
} else {
    figures[I] = ...
}
figures[I].draw()

**Review by olekskov — CHANGES_REQUESTED** [review]

**Pull request closed without merging**